bestbook/api/books.py

```diff
@@ -43,17 +46,59 @@ class Topic(core.Base):
                      nullable=False)
     modified = Column(DateTime(timezone=False), default=None)
 
+    @classmethod
+    def upsert(cls, topic):        
+        if isinstance(topic, cls):
+            return topic
+        try:
+            return cls.get(name=topic)
+        except:
+            return cls(name=topic).create()
 
 class Book(core.Base):
 
     __tablename__ = "books"
+    __table_args__ = (UniqueConstraint('work_olid', 'edition_olid', name='_work_edition_uc'),)
 
-    work_olid = Column(Unicode, nullable=False, unique=True, primary_key=True) # Open Library ID (required)
+    id = Column(BigInteger, primary_key=True)
+    work_olid = Column(Unicode, nullable=False) # Open Library ID (required)
     edition_olid = Column(Unicode, nullable=True) # Open Library ID (optional)
     created = Column(DateTime(timezone=False), default=datetime.utcnow,
                      nullable=False)
     modified = Column(DateTime(timezone=False), default=None)
 
+    @staticmethod
+    def clean_olid(olid):
+        """Extract just the olid from some olid-containing string"""
+        if olid.lower().startswith('ol') and olid.lower()[-1] in ['M', 'W']:
```

Comment:
```suggestion
        if olid.lower().startswith('ol') and olid.lower().endswith(('M', 'W')):
```
Easier to read and will not raise an IndexError if olid is an empty string.